File: test/kotoba/lang/pqh/signal_test.clj
(ns kotoba.lang.pqh.signal-test
  "Port of signal.test.ts (deprecated stand-in) + the pqh-v1 session section
   of pq.test.ts."
  (:require [clojure.test :refer [deftest is testing use-fixtures]]
            [kotoba.lang.pqh.signal :as sig]
            [kotoba.lang.pqh.pq :as pq]
            [kotoba.lang.pqh.crypto :as crypto]
            [kotoba.lang.pqh.aead-bc :as aead-bc]
            [kotoba.lang.pqh.pq-bc :as pq-bc]
            [kotoba.lang.pqh.util :as u]))

(use-fixtures :each (fn [f] (sig/clear-sessions!) (binding [crypto/*aead* (aead-bc/bc-aead) pq/*pq* (pq-bc/bc-pq)] (f))))

(def sender-did "did:example:sender")
(def recipient-did "did:example:recipient")

;; ── port of signal.test.ts ──────────────────────────────────────────────

(deftest deprecated-session-wrap-unwrap
  (testing "successful wrapKey/unwrapKey roundtrip"
    (let [session (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          plaintext "This is a secret message for key wrapping."
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext plaintext})]
      (is (= plaintext (sig/unwrap-key {:session session :ciphertext ciphertext})))))

  (testing "throws when unwrapping with a different session"
    (let [session1 (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          session2 (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          {:keys [ciphertext]} (sig/wrap-key {:session session1 :plaintext "This should not be decryptable."})]
      (is (thrown? Exception (sig/unwrap-key {:session session2 :ciphertext ciphertext})))))

  (testing "handles an empty plaintext string"
    (let [session (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext ""})]
      (is (= "" (sig/unwrap-key {:session session :ciphertext ciphertext})))))

  (testing "handles plaintext with unicode characters"
    (let [session (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          plaintext "こんにちは、世界！ (Hello, World!) 😃"
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext plaintext})]
      (is (= plaintext (sig/unwrap-key {:session session :ciphertext ciphertext})))))

  (testing "handles a large (16KB) plaintext"
    (let [session (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          large-string (u/bytes->hex (u/random-bytes (* 16 1024)))
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext large-string})]
      (is (= large-string (sig/unwrap-key {:session session :ciphertext ciphertext})))))

  (testing "throws when unwrapping with an invalid session handle"
    (let [session (sig/establish-session {:sender-did sender-did :recipient-did recipient-did})
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext "test"})]
      (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Invalid session handle"
                            (sig/unwrap-key {:session "invalid-session-handle" :ciphertext ciphertext})))))

  (testing "throws when wrapping with an invalid session handle"
    (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Invalid session handle"
                          (sig/wrap-key {:session "invalid-session-handle" :plaintext "test"})))))

;; ── port of pq.test.ts's "signal pqh-v1 sessions" ─────────────────────────

(deftest pqh-v1-sessions
  (testing "initiator and responder derive interoperable session keys"
    (let [kem (pq/generate-hybrid-kem-key-pair)
          {:keys [session handshake]} (sig/establish-session-initiator
                                         {:sender-did sender-did :recipient-did recipient-did
                                          :recipient-kem (:public-bundle kem)})
          responder (sig/establish-session-responder
                      {:sender-did sender-did :recipient-did recipient-did
                       :handshake handshake
                       :recipient-kem-secret (:secret-bundle kem)
                       :recipient-kem-public (:public-bundle kem)})
          plaintext "per-record symmetric key material"
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext plaintext})]
      (is (= plaintext (sig/unwrap-key {:session responder :ciphertext ciphertext})))))

  (testing "a responder with the wrong KEM secret cannot unwrap"
    (let [kem (pq/generate-hybrid-kem-key-pair)
          wrong (pq/generate-hybrid-kem-key-pair)
          {:keys [session handshake]} (sig/establish-session-initiator
                                         {:sender-did sender-did :recipient-did recipient-did
                                          :recipient-kem (:public-bundle kem)})
          responder (sig/establish-session-responder
                      {:sender-did sender-did :recipient-did recipient-did
                       :handshake handshake
                       :recipient-kem-secret (:secret-bundle wrong)
                       :recipient-kem-public (:public-bundle wrong)})
          {:keys [ciphertext]} (sig/wrap-key {:session session :plaintext "secret"})]
      (is (thrown? Exception (sig/unwrap-key {:session responder :ciphertext ciphertext}))))))

File: test/signal.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { establishSession, wrapKey, unwrapKey, _clearSessions } from '../src/signal';
import { randomBytes } from '@noble/hashes/utils';

describe('@etzhayyim/sdk: signal', () => {
  beforeEach(() => {
    _clearSessions();
  });

  const senderDid = 'did:example:sender';
  const recipientDid = 'did:example:recipient';

  it('should perform a successful wrapKey/unwrapKey roundtrip', () => {
    const session = establishSession({ senderDid, recipientDid });
    const plaintext = 'This is a secret message for key wrapping.';

    const { ciphertext } = wrapKey({ session, plaintext });
    const decrypted = unwrapKey({ session, ciphertext });

    expect(decrypted).toBe(plaintext);
  });

  it('should throw an error when unwrapping with a different session', () => {
    const session1 = establishSession({ senderDid, recipientDid });
    const session2 = establishSession({ senderDid, recipientDid });
    const plaintext = 'This should not be decryptable.';

    const { ciphertext } = wrapKey({ session: session1, plaintext });

    // Noble ciphers will throw on AEAD tag mismatch
    expect(() => {
      unwrapKey({ session: session2, ciphertext });
    }).toThrow();
  });

  it('should correctly handle an empty plaintext string', () => {
    const session = establishSession({ senderDid, recipientDid });
    const plaintext = '';

    const { ciphertext } = wrapKey({ session, plaintext });
    const decrypted = unwrapKey({ session, ciphertext });

    expect(decrypted).toBe(plaintext);
  });

  it('should correctly handle plaintext with unicode characters', () => {
    const session = establishSession({ senderDid, recipientDid });
    const plaintext = 'こんにちは、世界！ (Hello, World!) 😃';

    const { ciphertext } = wrapKey({ session, plaintext });
    const decrypted = unwrapKey({ session, ciphertext });

    expect(decrypted).toBe(plaintext);
  });

  it('should correctly handle a large (16KB) plaintext', () => {
    const session = establishSession({ senderDid, recipientDid });
    const largeString = Buffer.from(randomBytes(16 * 1024)).toString('base64');

    const { ciphertext } = wrapKey({ session, plaintext: largeString });
    const decrypted = unwrapKey({ session, ciphertext });

    expect(decrypted).toBe(largeString);
  });

  it('should throw when unwrapping with an invalid session handle', () => {
    const session = establishSession({ senderDid, recipientDid });
    const { ciphertext } = wrapKey({ session, plaintext: 'test' });

    expect(() => {
      unwrapKey({ session: 'invalid-session-handle', ciphertext });
    }).toThrow('Invalid session handle');
  });

  it('should throw when wrapping with an invalid session handle', () => {
    expect(() => {
      wrapKey({
        session: 'invalid-session-handle',
        plaintext: 'test',
      });
    }).toThrow('Invalid session handle');
  });
});

